### PR TITLE
feat(gateway): add event types

### PR DESCRIPTION
**Pull request by lahgolz**

### CODE REVIEW

File: src/clients/Client.ts [hunk 1]
```diff
@@ -1,8 +1,55 @@
+import { Snowflake } from 'discord-api-types';
 import merge from 'lodash.merge';
+import EventEmitter from 'node:events';
+import { CloseEvent } from 'ws';
 import { WebSocketManager } from '../gateway/WebSocketManager';
+import { Events } from '../utils/events';
```
Comment (Mesteery) [inline src/clients/Client.ts]: `Events` should not be in `utils`.

File: src/clients/Client.ts [hunk 2]
```diff
@@ -1,8 +1,55 @@
+import { Snowflake } from 'discord-api-types';
 import merge from 'lodash.merge';
+import EventEmitter from 'node:events';
```
Comment (Mesteery) [inline src/clients/Client.ts]:
```suggestion
import EventEmitter from 'events';
```

File: src/clients/Client.ts [hunk 3]
```diff
@@ -67,6 +114,7 @@ const defaultOptions: DeepRequired<ClientOptions> = {
 	},
 };
 
+//eslint-disable-next-line no-redeclare
```
Comment (Mesteery) [inline src/clients/Client.ts]:
```suggestion
// eslint-disable-next-line no-redeclare
```

File: src/gateway/WebSocketClient.ts [hunk 4]
```diff
@@ -260,15 +260,15 @@ export class WebSocketClient extends EventEmitter {
...
+			this.emit(WebSocketEvents.READY);
 
 			return;
 		}
 
 		this.readyTimeout = this.client.setTimeout(() => {
 			this.readyTimeout = undefined;
 			this.status = Status.READY;
-			// this.emit(ShardEvents.ALL_READY, this.expectedGuilds);
+			this.emit(WebSocketEvents.READY, this.expectedGuilds);
```
Comment (Mesteery) [inline src/gateway/WebSocketClient.ts:272]: why there is `this.expectedGuilds` as argument here ?

File: src/gateway/WebSocketManager.ts [hunk 5]
```diff
@@ -65,16 +68,17 @@ export class WebSocketManager extends EventEmitter {
...
 				this.reconnecting = false;
 				this.triggerClientReady();
 			});
 
-			this.webSocketClient.on(WebSocketEvents.CLOSE, (event) => {
+			this.webSocketClient.on(WebSocketEvents.CLOSE, (event: CloseEvent) => {
 				if (event.code === 1000 ? this.destroyed : UNRECOVERABLE_CLOSE_CODES.has(event.code)) {
-					// this.client.emit(Events.SHARD_DISCONNECT, event, this.webSocketClient.id);
+					this.client.emit(Events.WEBSOCKET_DISCONNECTING, event);
+
```
Comment (Mesteery) [inline src/gateway/WebSocketManager.ts]:
```suggestion
```

File: src/gateway/WebSocketManager.ts [hunk 6]
```diff
@@ -92,12 +96,13 @@ export class WebSocketManager extends EventEmitter {
...
-				this.client.emit(Events.SHARD_RECONNECTING, this.webSocketClient.id);
-			});*/
+			this.webSocketClient.on(WebSocketEvents.INVALID_SESSION, () => {
+				this.client.emit(Events.WEBSOCKET_RECONNECTING);
+			});
 
 			this.webSocketClient.on(WebSocketEvents.DESTROYED, () => {
-				//this.client.emit(Events.SHARD_RECONNECTING, this.webSocketClient.id);
+				this.client.emit(Events.WEBSOCKET_RECONNECTING);
+
```
Comment (Mesteery) [inline src/gateway/WebSocketManager.ts]:
```suggestion
```

File: src/utils/events.ts [hunk 7]
```diff
@@ -15,3 +15,12 @@ export function once(emitter: EventEmitter, name: string, signal: unknown): Prom
 export function rejectOnce(emitter: EventEmitter, name: string, signal: unknown): Promise<any> {
 	return once(emitter, name, signal).then(Promise.reject);
 }
+
+export enum Events {
```
Comment (Mesteery) [inline src/utils/events.ts]: this is not appriopriated here.

File: src/clients/Client.ts [hunk 8]
```diff
@@ -1,8 +1,55 @@
...
+import EventEmitter from 'node:events';
+import { CloseEvent } from 'ws';
 import { WebSocketManager } from '../gateway/WebSocketManager';
+import { Events } from '../utils/events';
 import { DeepRequired } from '../utils/types';
 import { BaseClient, BaseClientOptions } from './BaseClient';
 
+interface ClientEvents extends Record<Events, readonly unknown[]> {
+	ready: [];
+	websocketDisconnecting: [CloseEvent];
```
Comment (Mesteery) [inline src/clients/Client.ts]:
```suggestion
	websocketDisconnect: [CloseEvent];
```
it's maybe a better name ?
(`gatewayDisconnect` ?)
same for others

**Review by Mesteery — CHANGES_REQUESTED** [review]

File: src/gateway/WebSocketClient.ts (hunk 4 above)
Comment (lahgolz) [reply src/gateway/WebSocketClient.ts:272]: See https://discord.com/developers/docs/topics/gateway#ready

File: src/utils/events.ts (hunk 7 above)
Comment (lahgolz) [reply src/utils/events.ts]: Where should I put this?

File: src/gateway/WebSocketClient.ts (hunk 4 above)
Comment (lahgolz) [reply src/gateway/WebSocketClient.ts:272]: But it's something specific to the sharding system, so I'm not sure...

File: src/utils/events.ts (hunk 7 above)
Comment (Mesteery) [reply src/utils/events.ts]: Maybe in gateway folder ? In index ?

File: src/gateway/Events.ts [hunk 9]
```diff
@@ -0,0 +1,8 @@
+export enum Events {
+	CLIENT_READY = 'ready',
+	WEBSOCKET_DISCONNECTING = 'websocketDisconnecting',
```
Comment (Mesteery) [inline src/gateway/Events.ts]: GATEWAY_DISCONNECTION

File: src/gateway/Events.ts [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+	WEBSOCKET_ERROR = 'websocketError',
+	WEBSOCKET_READY = 'websocketReady',
+	WEBSOCKET_RECONNECTING = 'websocketReconnecting',
+	WEBSOCKET_RECONNECTINGf = 'websocketReconnecting',
```
Comment (Mesteery) [inline src/gateway/Events.ts]:
```suggestion
```

File: src/gateway/Events.ts [hunk 11]
```diff
@@ -0,0 +1,8 @@
+export enum Events {
+	CLIENT_READY = 'ready',
+	WEBSOCKET_DISCONNECTING = 'websocketDisconnecting',
+	WEBSOCKET_ERROR = 'websocketError',
+	WEBSOCKET_READY = 'websocketReady',
+	WEBSOCKET_RECONNECTING = 'websocketReconnecting',
```
Comment (Mesteery) [inline src/gateway/Events.ts]:
```suggestion
	GATEWAY_RECONNECTION = 'gatewayReconnection',
```

File: src/gateway/Events.ts [hunk 12]
```diff
@@ -0,0 +1,8 @@
+export enum Events {
+	CLIENT_READY = 'ready',
+	WEBSOCKET_DISCONNECTING = 'websocketDisconnecting',
+	WEBSOCKET_ERROR = 'websocketError',
+	WEBSOCKET_READY = 'websocketReady',
```
Comment (Mesteery) [inline src/gateway/Events.ts]:
```suggestion
	GATEWAY_READY = 'gatewayReady',
```

File: src/gateway/Events.ts (hunk 9 above)
Comment (Mesteery) [reply src/gateway/Events.ts]: or WEBSOCKET/webSocket (or websocket) (same for below/above)

**Review by Mesteery — CHANGES_REQUESTED** [review]

File: src/gateway/Events.ts [hunk 13]
```diff
@@ -0,0 +1,7 @@
+export enum Events {
+	CLIENT_READY = 'ready',
+	WEBSOCKET_DISCONNECTING = 'websocketDisconnecting',
```
Comment (Mesteery) [inline src/gateway/Events.ts]:
```suggestion
	WEBSOCKET_DISCONNECTINON = 'websocketDisconnection',
```

File: src/gateway/Events.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
...
+	WEBSOCKET_ERROR = 'websocketError',
```
Comment (Mesteery) [inline src/gateway/Events.ts]: gateway or websocket (this should be consistent with other events name)

File: src/gateway/Events.ts [hunk 15]
```diff
@@ -0,0 +1,7 @@
+export enum Events {
```
Comment (Mesteery) [inline src/gateway/Events.ts]:
```suggestion
export const enum Events {
```

**Review by Mesteery — CHANGES_REQUESTED** [review]

**Review by Mesteery — APPROVED** [review]

**Review by Aakodal — APPROVED** [review]

**Pull request merged**